From 5be3c3ef119402d805d7b3fccd6b79dfa5249114 Mon Sep 17 00:00:00 2001
From: Guillaume <guillaume.thouvenin@vates.tech>
Date: Mon, 31 Mar 2025 14:59:05 +0200
Subject: [PATCH] [xcp-ng] fix IPv6 import

Add brackets in `Host` header of http request when relevant.
See: https://github.com/mirage/ocaml-cohttp/issues/997

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>
---
 ocaml/xapi/xapi_vm.ml | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/ocaml/xapi/xapi_vm.ml b/ocaml/xapi/xapi_vm.ml
index 2d4595b6d..a5b94858f 100644
--- a/ocaml/xapi/xapi_vm.ml
+++ b/ocaml/xapi/xapi_vm.ml
@@ -1562,7 +1562,23 @@ let rec import_inner n ~__context ~url ~sr ~full_restore ~force =
             Cohttp.Request.Make (Cohttp_posix_io.Unbuffered_IO) in
           let module Response =
             Cohttp.Response.Make (Cohttp_posix_io.Unbuffered_IO) in
-          let request = Cohttp.Request.make ~meth:`GET uri in
+          let maybe_wrap_IPv6_literal addr =
+            if Unixext.domain_of_addr addr = Some Unix.PF_INET6 then
+              "[" ^ addr ^ "]"
+            else
+              addr
+          in
+          let headers = Cohttp.Header.init () in
+          let headers =
+            Cohttp.Header.add_unless_exists headers "host"
+              (match Uri.scheme uri with
+              | Some "httpunix" -> ""
+              | _ -> (
+                (maybe_wrap_IPv6_literal (Uri.host_with_default ~default:"localhost" uri))
+                  ^
+                  match Uri.port uri with Some p -> ":" ^ string_of_int p | None -> ""))
+          in
+          let request = Cohttp.Request.make ~meth:`GET ?headers:(Some headers) uri in
           let ic =
             {
               Cohttp_posix_io.Unbuffered_IO.header_buffer= None
